### PR TITLE
Implement Prismarine Auth

### PR DESCRIPTION
**Pull request by kashalls**
Removes authentication handling from the protocol and moves it to a joint module `prismarine-auth`.

### CODE REVIEW

**Comment by kashalls** [comment]
I can confirm that both microsoft device code and xbox password authentication is working! Please do double test this as I could be stupid and made a mistake somewhere. 

Password Authentication Here;
image

We do not cache the mba token because of https://github.com/PrismarineJS/prismarine-auth/blob/master/src/MicrosoftAuthFlow.js#L163
If we can get a consensus about that, I can implement that fix real quick. 

Ready for review!

File: src/client/auth.js [hunk 1]
```diff
@@ -1,31 +1,44 @@
...
-  // This profile / session here could be simplified down to where it just passes the uuid of the player to encrypt.js
-  // That way you could remove some lines of code. It accesses client.session.selectedProfile.id so /shrug.
-  // - Kashalls
-  const profile = {
-    name: xboxProfile?.extraData?.displayName || 'Player',
-    uuid: xboxProfile?.extraData?.identity || 'adfcf5ca-206c-404a-aec4-f59fff264c9b', // random
-    xuid: xboxProfile?.extraData?.XUID || 0
+async function authenticate (client, options) {
+  if (!options.profilesFolder) {
+    options.profilesFolder = minecraftFolderPath
```
Comment (extremeheat) [inline src/client/auth.js]: Should be minecraftFolderPath + '/nmp-cache'

Comment (kashalls) [reply src/client/auth.js]: I should probably implement this fix into nmp as well.

File: src/client/auth.js [hunk 2]
```diff
@@ -38,49 +51,19 @@ function createOfflineSession (client, options) {
...
     xuid: 0
   }
+  return postAuthenticate(client, profile, []) // No extra JWTs, only send 1 client signed chain with all the data
+}
+
+function postAuthenticate (client, profile, chains) {
   client.profile = profile
   client.username = profile.name
-  client.accessToken = [] // No extra JWTs, only send 1 client signed chain with all the data
+  client.accessToken = chains // No extra JWTs, only send 1 client signed chain with all the data
```
Comment (extremeheat) [inline src/client/auth.js]: you can remove the comment, it's only for offline mode

**Review by rom1504 — APPROVED** [review]

**Pull request merged**